design/physical-network-api-prototype.md

```diff
@@ -0,0 +1,176 @@
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+The Metal³ project is currently centered around an API for managing physical
+hosts.  This proposal is to explore expanding the scope to also include an API
+for managing physical network devices.  This exploration would be done by
+starting with a prototype that can configure some aspects of a top-of-rack
+(ToR) switch.
+
+## Motivation
+
+Metal³ follows the paradigm of Kubernetes Native Infrastructure (KNI), which is
+an approach to use Kubernetes to manage underlying infrastructure.  Managing
+the configuration of some physical network devices is closely related to
+managing physical hosts.
+
+As bare metal hosts are provisioned or later repurposed, there may be
+corresponding physical network changes that must be made, such as reconfiguring
+a ToR switch port.  If the provisioning of the physical host is managed through
+a Kubernetes API, it would be convenient to be able to reconfigure related
+network devices using a similar API.
+
+### Goals
+
+Produce a `physical-network-api` prototype, which includes the following:
+
+* A `Switch` CRD with one or more fields that corresponding to switch
+  configuration, ideally with a vendor neutral definition
+* A demonstration of using this prototype API to configure a switch
+* A documented retrospective that includes lessons learned and proposed next
+  steps
+* Evaluate the re-use of a single existing network device configuration
+  technology
+
+### Non-Goals
+
+Out of scope:
+
+* Testing with anything more than a single switch model
+* Prototyping an API for anything beyond a Switch
+* Full configuration possibilities of a switch
+* Evaluation of all potential technologies that could be re-used for network
+  device configuration.  Future prototypes can be done to consider other
+  alternatives if desired.
+
+These items may become in scope after a prototype is reviewed and next steps
+discussed.
+
+## Proposal
+
+The prototype should explore the re-use of [Ansible
+Networking](https://docs.ansible.com/ansible/latest/network/index.html) to
+perform device configuration.  Ansible includes modules for managing the
+configuration of many different network devices.  Creation of CRDs in this new
+API can use modeling from Ansible as inspiration, particularly if there is any
+vendor neutral modeling already done there.
+
+Another project that could be used as inspiration is the [Ansible Networking
+Neutron ML2 Driver](https://networking-ansible.readthedocs.io/en/latest/) which
+created a Python API to abstract generic switch configuration on top of Anisble
+networking modules.
+
+The [operator-sdk](https://github.com/operator-framework/operator-sdk) project
+includes some support for Ansible operators.  This should be considered as one
+approach for re-use of Ansible networking modules.  Other approaches to Ansible
+networking re-use can be considered and discussed in the read-out from the
+prototype.
```

Comment:
What network-runner provides is a series of Ansible roles with cross-platform implementations, plus a Python library with an API that internally creates a playbook to run a role. So the two most likely options are probably:

* generate playbooks for each role and run them using an Ansible Operator; or
* add a gRPC API to network-runner and call it from a Golang Operator.

Reply:
Thanks!  I'll incorporate the ideas.